docs/design-notes/per-partition-rate-limit.md

```diff
@@ -0,0 +1,229 @@
+# Per-partition rate limiting
+
+Scylla clusters operate best when the data is spread across a large number
+of small partitions, and reads/writes are spread uniformly across all shards
+and nodes. Due to various reasons (bugs, malicious end users etc.) this
+assumption may suddenly not hold anymore and one partition may start getting
+a disproportionate number of requests. In turn, this usually leads to the owning
+shards being overloaded - a scenario called "hot partition" - and the total
+cluster latency becoming worse.
+
+The _per partition rate limit_ feature allows users to limit the rate
+of accepted requests on a per-partition basis. When a partition exceeds
+the configured limit of operations of given type (reads/writes) per second,
+the cluster will start responding with errors to some of the operations for that
+partition so that, statistically, the rate of accepted requests is kept
+at the configured limit. Rejected operations use less resources, therefore
+this feature can help in the "hot partition" situation.
+
+_NOTE_: this is an overload protection mechanism and may not be used to reliably
+enforce limits in some situations. Due to Scylla's distributed nature,
+the actual number of accepted requests depends on the cluster and driver
+configuration and may be larger by a factor of RF (keyspace's rpelication
+factor). It is recommended to set the limit to a value an order of magnitude
+larger than the maximum expected per-partition throughput.
+
+## Usage
+
+### Server-side configuration
+
+Per-partition limits are set separately for reads and writes, on a per-table
+basis. Limits can be set with the `per_partition_rate_limit` extension when
+CREATE'ing or ALTER'ing a table using a schema extension:
+
+```cql
+ALTER TABLE ks.tbl WITH per_partition_rate_limit = {
+    'max_reads_per_second': 123,
+    'max_writes_per_second': 456
+};
+```
+
+Both `max_reads_per_second` and `max_writes_per_second` are optional - omitting
+one of them means "no limit" for that type of operation.
+
+### Driver response
+
+Rejected operations are reported as an ERROR response to the driver.
+If the driver supports it, the response contains a scylla-specific error code
+indicating that the operation was rejected. For more details about the error
+code, see the [Rate limit error](./protocol-extensions.md#Rate%20limit%20error)
+section in the `protocol-extensions.md` doc.
+
+If the driver doesn't support the new error code, the `Config_error` code
+is returned instead. The code was chosen in order for the retry policies
+of the drivers not to retry the requests and instead propagate them directly
+to the users.
+
+## Implementation
+
+Per-partition limits are tracked and enforced by replicas. When the rate
+of requests to a partition exceed configured limit, they start being rejected
+with such a probability that, statistically, the number of accepted reject
+falls back to the configured limit.
+
```

Comment:
> I see problems with rejecting on replicas. First of all it may causes a lot of work by a coordinator to be wasted. For writes we send to all replicas, so if some reject we may still complete the operation successfully, but for reads we contact only CL with probably one replica to retry with and the chance for operation to fail becomes large while a coordinator already wasted a lot of effort.

Yes, when the coordinator is not a replica replicas will do some wasted work. However, even still less work is actually done compared to not rate limiting at all because replicas sometimes will reject - so the total work the replica shards have to do will be lower than without rate limiting. I'm not sure how it could be done better - in this design, replicas are responsible for tracking partitions that they own, so letting replicas decide avoids the need for additional round trips.

Note that this problem does not happen for shard-aware drivers as in this case the coordinator usually _is_ a replica and it is allowed to decide by itself whether to reject or not. In this case, the coordinator either proceeds with the operation and replicas are not allowed to reject it, or the coordinator rejects the operation and does not contact the replicas.

> 
> Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.

A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.

Reply:
> > I see problems with rejecting on replicas. First of all it may causes a lot of work by a coordinator to be wasted. For writes we send to all replicas, so if some reject we may still complete the operation successfully, but for reads we contact only CL with probably one replica to retry with and the chance for operation to fail becomes large while a coordinator already wasted a lot of effort.
> 
> Yes, when the coordinator is not a replica replicas will do some wasted work. However, even still less work is actually done compared to not rate limiting at all because replicas sometimes will reject - so the total work the replica shards have to do will be lower than without rate limiting. I'm not sure how it could be done better - in this design, replicas are responsible for tracking partitions that they own, so letting replicas decide avoids the need for additional round trips.

As I said in some other comment I am not so sure non shard aware driver is a case worth covering especially as the solution is far from been perfect. @avikivity?

> 
> Note that this problem does not happen for shard-aware drivers as in this case the coordinator usually _is_ a replica and it is allowed to decide by itself whether to reject or not. In this case, the coordinator either proceeds with the operation and replicas are not allowed to reject it, or the coordinator rejects the operation and does not contact the replicas.
> 
> > Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.
> 
> A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.

So in that case you actually multiplied the work you are doing. You add load to the coordinator and do not drop any load from replicas since hints will be replayed shortly. I cannot say how rare on otherwise this case is (given that probabilities are involved in each replica's decision).

Reply:
> > > Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.
> > 
> > 
> > A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.
> 
> So in that case you actually multiplied the work you are doing. You add load to the coordinator and do not drop any load from replicas since hints will be replayed shortly. I cannot say how rare on otherwise this case is (given that probabilities are involved in each replica's decision).

Yes, I do more work for accepted writes, but I also do less work for rejected writes. We do not write hints for failed writes. For example, if the limit is 100 ops/s for a partition and a client starts sending 10k ops/s, with CL=QUORUM and RF=3 we will generate 100 hints for 100 successful writes but we won't generate any for the remaining 9900. The overall load will be lower.

One thing I should probably mention which might be not obvious here: if a throughput for a partition is kept below the limit, then replicas will never reject. No hints will be generated. We do not pay the cost of hints if the user behaves nicely.

There is probably a range between `limit` and `limit * k` where `k` is a small constant larger than one, such that the load is higher than if rate limiting was disabled. I'm not sure if it is a big problem - the limit should be a small fraction of what the shard can handle (otherwise there won't be much resources left for other shards when rate limiting kicks in). I would expect that in the case of a "hot partition" problem, the number of overloaded partitions is small.

Reply:
> > > > Also what happens if some replicas rejects a write but it still succeeds? Do we write hints in this case? Do we reply hints with writes that can be rejected? Because if not we achieved nothing, just delayed the write a little.
> > > 
> > > 
> > > A write is considered failed if less then `CL` replicas accept. We write hints only for writes which succeed. So, if `L` ops/s is the limit, then we will write at most `L * (RF - CL)` additional hints/s. Again, those hints will only be generated when _not_ using a shard-aware driver.
> > 
> > 
> > So in that case you actually multiplied the work you are doing. You add load to the coordinator and do not drop any load from replicas since hints will be replayed shortly. I cannot say how rare on otherwise this case is (given that probabilities are involved in each replica's decision).
> 
> Yes, I do more work for accepted writes, but I also do less work for rejected writes. We do not write hints for failed writes. For example, if the limit is 100 ops/s for a partition and a client starts sending 10k ops/s, with CL=QUORUM and RF=3 we will generate 100 hints for 100 successful writes but we won't generate any for the remaining 9900. The overall load will be lower.

I am not discussing a failed writes here. I am talking about writes that were suppose to fail because one of the replicas is overloaded (if the write would gi through that replica it would have failed) and asks to rate limit access, but because we "distributed" decision about the failure the write may actually succeed and overload the overloaded replica and the coordinator. Note that the issue is not only hint that is written but also write that happens when hint is replayed. So the replica tries to avoid the write, but didn't.  


> 
> One thing I should probably mention which might be not obvious here: if a throughput for a partition is kept below the limit, then replicas will never reject. No hints will be generated. We do not pay the cost of hints if the user behaves nicely.

This is obvious.

Reply:
> > > I see problems with rejecting on replicas. First of all it may causes a lot of work by a coordinator to be wasted. For writes we send to all replicas, so if some reject we may still complete the operation successfully, but for reads we contact only CL with probably one replica to retry with and the chance for operation to fail becomes large while a coordinator already wasted a lot of effort.
> > 
> > 
> > Yes, when the coordinator is not a replica replicas will do some wasted work. However, even still less work is actually done compared to not rate limiting at all because replicas sometimes will reject - so the total work the replica shards have to do will be lower than without rate limiting. I'm not sure how it could be done better - in this design, replicas are responsible for tracking partitions that they own, so letting replicas decide avoids the need for additional round trips.
> 
> As I said in some other comment I am not so sure non shard aware driver is a case worth covering especially as the solution is far from been perfect. @avikivity?

I think that we do care about users which use non-shard-aware drivers.

Now that I think about it, an idea comes to my mind how to avoid waste: replicas could inform coordinators about the level of overload given partition has (a bit like the view backlog). This information would be sent only when the partition is overloaded and coordinators would keep it in a small cache which expires over time. Then, a coordinator could use information in its local cache to probabilistically reject. The main problems seem to be choosing the cache size and maybe that the information in the cache can be a bit stale which could affect the accept rates.

Of course, this would require changing this PR quite a bit and seeing if this idea works alright (I would estimate 2-4 days of work), so this should also be considered while deciding whether to do it now, later, or at all.

WDYT @avikivity @gleb-cloudius ? Any other ideas?

Reply:
For the majority of queries, won't all replicas reach the same decision? For example, if the rate limit is 100/s and people are sending us 1000 queries/s. then for some queries we may get mismatched responses but for at least 700 all replicas will fail?

Reply:
The probability is not entirely trivial to compuate, so I ran an experiment: RF=3, CL=ALL, limit 100/s, workload 1000 reads/s. Each replica reported a bit above 400 successful reads/s.

- 59% cases had 0 successes
- 21% cases had 1 success
- 6% cases had 2 successes
- 14% cases had 3 successes (higher than 10% because of the coordinator optimization)

So, in about 59% + 14% = 73% cases all participating replicas made the same decision.

Reply:
I am super pleased that my 70% guess was so close to reality.

In the future we can try to coordinate the failures. If the algorithm is probabilistic, we can generate a random number on the coordinator and send it to the replicas, so they'll make similar decisions if their states are similar.

Reply:
> The probability is not entirely trivial to compuate, so I ran an experiment: RF=3, CL=ALL, limit 100/s, workload 1000 reads/s. Each replica reported a bit above 400 successful reads/s.
> 
>     * 59% cases had 0 successes
> 
>     * 21% cases had 1 success
> 
>     * 6% cases had 2 successes
> 
>     * 14% cases had 3 successes (higher than 10% because of the coordinator optimization)
> 
> 
> So, in about 59% + 14% = 73% cases all participating replicas made the same decision.

In real life with more networking delays and node restarts (loosing statistics) we may have much more cases where replicas disagree. Even in this simple test there are 6% requests that will increase local on a system instead of decreasing it.

Reply:
Why do the 6% cases increase load? They decreased it, though not as much as they could have.

Reply:
btw, I think that the coordinator decision should account for RF (write) or CL (read) attempts, not 1, since that's the amount of work it saved. That balances out whether we had a coordinator or replica decision.

Reply:
Ah, that's why account_only and account_and_enforce. But maybe RF-accounting at the coordinator is simpler.

Reply:
> Why do the 6% cases increase load? They decreased it, though not as much as they could have.

In those 6% of the cases a write query will succeed and thus for one replica that replied with an error a hint will be stored. Said hint will be replied shortly (since the node is alive), but this time with forced apply. This result in more work, not less for non trivial amount of requests even in a simple test.

Reply:
Yes. Well it's unfortunate. We should try to get replicas to arrive at the same decision (by sending them the random number for a decision), but we can do it post-merge.

Reply:
I implemented the idea with the coordinator sending a random number to replicas. As a result, probability calculation is now much simpler.

Reply:
> I'll test cluster upgrade tomorrow. I also measured a regression of 600 insns/op for writes and 500 for reads, I'll investigate where it comes from and see what can be done about it.

First, check if there is a regression in the number of tasks/allocations.